extra/changelog-generator/changelog-generator

```diff
@@ -139,6 +146,73 @@ parser.add_argument(
     + "Note that for technical reasons these must come last.",
 )
 
+
+class GitQuerier:
+    def __init__(self, gitargs, base_dir):
+        self.gitargs = gitargs
+        self.base_dir = base_dir
+
+    def get_commits_for_range(self, repo, range):
+        output = subprocess.check_output(
+            ["git", "rev-list", "--reverse", range] + self.gitargs,
+            cwd=os.path.join(self.base_dir, repo),
+        )
+        return output.decode().split()
+
+    def get_raw_commit_message(self, repo, sha):
+        output = subprocess.check_output(
+            ["git", "cat-file", "-p", sha] + self.gitargs,
+            cwd=os.path.join(self.base_dir, repo),
+        )
+        # Return only the part after headers (double newline).
+        return output.decode().split("\n\n", 1)[1]
+
+    def show_commit_without_diff(self, repo, sha):
+        output = subprocess.check_output(
+            ["git", "show", "-s", sha] + self.gitargs,
+            cwd=os.path.join(self.base_dir, repo),
+        )
+        return output.decode()
+
+
+class GitHubQuerier:
```

Comment:
I was asking myself the exact same question, and I wanted to do it. But interfaces are kind of useless in Python, they give you nothing:

1. Because of duck typing it makes no difference when you have a variable referring to it.
2. Because there is no real interface/abstract type, but just superclass functions that do nothing, it cannot even warn you if you forget to implement something.

So in the end I decided it's not helpful and omitted it instead. I could include it for documentation purposes of course, but I thought it wasn't really worth it, and it could go quickly out of date too.

Reply:
It's fine :) 

There is the `abc` package  (abstract base class). But it's not important.

Reply:
Alright then.

But TIL, I didn't know about `abc`. I'll keep that in mind for next time.

Reply:
FWIW, the Python way of specifying interfaces is using protocol classes. See https://docs.python.org/3/library/typing.html#typing.Protocol

Reply:
I think `Protocol` is different, it is for specifying static type checks. Also useful, but independent of abstract base classes, and I think a bit overkill for this PR.

In the end I used `abc` to make an abstract base class.